File: src/Controllers/Web/TaskWebController.php
<?php

declare(strict_types=1);

namespace Taskmaster\Controllers\Web;

use InvalidArgumentException;
use Taskmaster\Core\Request;
use Taskmaster\Core\Response;
use Taskmaster\Services\TaskService;
use Taskmaster\Views\TaskWebView;
use Taskmaster\Models\Task;

final class TaskWebController {
   public function __construct(
      private readonly TaskService $taskService
   ) {
   }

   public function index(Request $request, array $params): void {
      $tasksData = $this->taskService->getTasks();
      $tasks = array_map(
         fn(array $data) => Task::fromArray($data),
         $tasksData
      );

      header('Content-Type: text/html; charset=utf-8');
      TaskWebView::index($tasks);
   }

   public function create(Request $request, array $params): void {
      header('Content-Type: text/html; charset=utf-8');
      TaskWebView::create();
   }

   public function store(Request $request, array $params): void {
      try {
         $this->taskService->createTask($request->body);
         header('Location: /tasks');
         exit;
      } catch (InvalidArgumentException $e) {
         http_response_code(422);
         header('Content-Type: text/html; charset=utf-8');
         echo "Validation Error: " . htmlspecialchars($e->getMessage()) . '<br><a href="/tasks/create">Back</a>';
         exit;
      }
   }

   public function edit(Request $request, array $params): void {
      $taskData = $this->taskService->getTask($params['id']);
      if ($taskData === null) {
         http_response_code(404);
         header('Content-Type: text/html; charset=utf-8');
         echo "Task not found";
         exit;
      }
      
      $task = Task::fromArray($taskData);
      header('Content-Type: text/html; charset=utf-8');
      TaskWebView::edit($task);
   }

   public function update(Request $request, array $params): void {
      try {
         $task = $this->taskService->updateTask(
            id: $params['id'],
            data: $request->body
         );
         if ($task === null) {
            http_response_code(404);
            header('Content-Type: text/html; charset=utf-8');
            echo "Task not found";
            exit;
         }
         header('Location: /tasks');
         exit;
      } catch (InvalidArgumentException $e) {
         http_response_code(422);
         header('Content-Type: text/html; charset=utf-8');
         echo "Validation Error: " . htmlspecialchars($e->getMessage()) . '<br><a href="/tasks/' . htmlspecialchars($params['id']) . '/edit">Back</a>';
         exit;
      }
   }

   public function destroy(Request $request, array $params): void {
      $deleted = $this->taskService->deleteTask($params['id']);
      if (!$deleted) {
         http_response_code(404);
         header('Content-Type: text/html; charset=utf-8');
         echo "Task not found";
         exit;
      }
      header('Location: /tasks');
      exit;
   }
}

File: src/main/java/com/taskmaster/controller/TaskWebController.java
package com.taskmaster.controller;

import com.taskmaster.model.Task;
import com.taskmaster.model.TaskStatus;
import com.taskmaster.service.TaskService;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Optional;

@Controller
@RequestMapping("/tasks")
public class TaskWebController {

    private final TaskService taskService;

    public TaskWebController(TaskService taskService) {
        this.taskService = taskService;
    }

    @GetMapping
    public String index(Model model) {
        List<Task> tasks = taskService.getTasks();
        model.addAttribute("tasks", tasks);
        return "tasks/list";
    }

    @GetMapping("/create")
    public String create(Model model) {
        model.addAttribute("statuses", TaskStatus.values());
        return "tasks/create";
    }

    @PostMapping
    public String store(
            @RequestParam String title,
            @RequestParam(required = false) String description,
            @RequestParam(required = false, defaultValue = "pending") String status,
            @RequestParam(required = false, name = "due_date") String dueDate,
            Model model) {
        try {
            taskService.createTask(title, description, status, dueDate);
            return "redirect:/tasks";
        } catch (IllegalArgumentException e) {
            model.addAttribute("errorMessage", e.getMessage());
            model.addAttribute("backUrl", "/tasks/create");
            model.addAttribute("statuses", TaskStatus.values());
            return "tasks/validation_error";
        }
    }

    @GetMapping("/{id}/edit")
    public String edit(@PathVariable String id, Model model) {
        Optional<Task> task = taskService.getTask(id);
        if (task.isEmpty()) {
            return "error/404";
        }
        model.addAttribute("task", task.get());
        model.addAttribute("statuses", TaskStatus.values());
        return "tasks/edit";
    }

    @PostMapping("/{id}/update")
    public String update(
            @PathVariable String id,
            @RequestParam String title,
            @RequestParam(required = false) String description,
            @RequestParam String status,
            @RequestParam(required = false, name = "due_date") String dueDate,
            Model model) {
        try {
            Optional<Task> updated = taskService.updateTask(id, title, description, status, dueDate);
            if (updated.isEmpty()) {
                return "error/404";
            }
            return "redirect:/tasks";
        } catch (IllegalArgumentException e) {
            model.addAttribute("errorMessage", e.getMessage());
            model.addAttribute("backUrl", "/tasks/" + id + "/edit");
            model.addAttribute("statuses", TaskStatus.values());
            return "tasks/validation_error";
        }
    }

    @PostMapping("/{id}/delete")
    public ResponseEntity<?> destroy(
            @PathVariable String id,
            @RequestHeader(value = "HX-Request", required = false) String hxRequest) {
        boolean deleted = taskService.deleteTask(id);
        if (!deleted) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body("Task not found");
        }
        if (hxRequest != null) {
            return ResponseEntity.ok("");
        }
        return ResponseEntity.status(HttpStatus.FOUND)
                .header("Location", "/tasks")
                .build();
    }
}
